### DISCUSSION TITLE
Open Roberta Lab Artificial Neural Networks for EV3

### DISCUSSION THREAD

**Discussion by Growflavor** [ev3dev/ev3dev#1646] [Show and tell]
Following up on #1645, creating this thread to discuss & share tinkering with EV3 and Artificial Neural Networks:
<img src="https://github.com/user-attachments/assets/5833921d-a693-407a-9bdb-5893e6a094c4" width="200">

**Direct link to Open Roberta Lab xNN for EV3:**
https://lab.open-roberta.org/?loadSystem=ev3dev&extensions=nn

links to tutorials: https://github.com/ev3dev/ev3dev/issues/1645#issuecomment-2294575138

Since the images in the tutorial PDFs are in German, I've shared templates for some of the example xNN shown in the tutorial PDFs via the Open Roberta Lab Gallery:

<img src="https://github.com/user-attachments/assets/0ba2dbd3-89d6-4616-8260-f28297a5d404" width="700">

Please share your EV3 xNN ideas & projects as well :-)

__________________________________________________________
[Open Roberta xNN is funded by the Ministry of School and Education of the State of North Rhine-Westphalia and developed in cooperation with teachers](https://www.iais.fraunhofer.de/de/presse/presseinformationen/presseinformationen-2022/presseinformation-22mmtt2.html)

**Comment by Growflavor** [comment]
I have found that Windows 10 & 11 do both support PC internet sharing with the EV3 via Bluetooth -- note the 192.168.138.nnn range IP in the screen snips below.
 
Consequently, a connection to Open Roberta Lab's EV3 programming interface www page via Bluetooth is possible...

In other words, with Windows 10 & 11 OpenRobertaLab may be used via the EV3 onboard Bluetooth,

making optional a purchase of a [USB WIFI adapter that works with EV3dev & EV3G such as the "Netis WF2120 Wireless N150 Nano USB Dongle, Ideal for Raspberry, Linux, RTL8188CUS"](https://www.amazon.com/dp/B008O2AL0K) :

**EV3 Bluetooth to PC Mobile Hotspot Connection via Windows 11 (click images to enlarge):**

<img src="https://github.com/user-attachments/assets/0255a2ce-5446-4b58-9493-874bd4874cf8" width="400">

<img src="https://github.com/user-attachments/assets/2c7e23a3-94e6-457b-91d8-b01dc9f764ce" width="400">


**EV3 Bluetooth to PC Mobile Hotspot Connection via Windows 10 (click images to enlarge):**

<img src="https://github.com/user-attachments/assets/61785034-ab22-488e-bf97-5d7d00a70832" width="500">

**Comment by Growflavor** [comment]
A first simple ANN color sensor drive speed controller (untrained so basically just a simple math function not much different than a P controller) based on the Open Roberta Lab tutorial (click images to enlarge):

<img src="https://github.com/user-attachments/assets/7bc761f6-4d3b-46cd-abaa-620cada0ac97" width="400">

<img src="https://github.com/user-attachments/assets/ac0eb193-5c74-4837-9c11-4533a7670958" width="400">
<img src="https://github.com/user-attachments/assets/38b9fd02-bc64-40e9-b14e-b53c8bf74d3d" width="400">
<img src="https://github.com/user-attachments/assets/d53eea51-dd30-47fa-ab00-38cfd6e43bb3" width="400">

**Comment by Growflavor** [comment]
An Arduino project using a trained ANN...

A) we 'ported' the Arduino "RoachBot" ;-) by @seanhodgins to the EV3 drive base with  4 color sensors & the Open Roberta Lab xNN interface:

https://www.instructables.com/Arduino-Neural-Network-Robot/

https://youtu.be/97R3TcUh5eI?si=P7N_xfIN4DT-XDgf&t=658

https://github.com/IdleHandsProject/makennbot?tab=readme-ov-file
https://github.com/JonComo/Arduino_weight_trainer

[giant_neural_network, the author of the above @JonComo "Arduino Weight Trainer", has an intro ANN youtube playlist](https://youtube.com/playlist?list=PLxt59R_fWVzT9bDxA76AHm3ig0Gg9S3So&si=P2OZjXGmdMHVrXy_)

<img src="https://github.com/user-attachments/assets/46046962-fd4a-4f16-a34b-9a3577c8a2ae" width="800">


B) Might also port to a Line Follower for comparison with P & PID:

**nikodembartnik**: "I am thinking about implementing neural net in a line follower project. I know that PID algorithm can be simpler faster and better but just for scientific purposes and because of my curiosity I want to try that.

**SeanHodgins**: "That is an excellent idea. This code would adapt well to that too if you wanted to try and port it over to your robot. This robot is basically an upside down line follower that follows shadows instead of lines."

see EV3 Line Follower P & PID formulas & discussion at: 
https://thecodingfun.com/2020/06/16/lego-mindstorms-ev3-pid-line-follower-code-by-using-micropython-2-0/

________________________________________________________________________
**Why might one explore this way with Open Roberta Lab's xNN?**  

--Reinforce understanding & study of differences in P, PID, & ANN formulas/methods...

--maybe look up "Karnaugh map":
**dimtass**: "For this case you could just use a lookup table which much faster than nested ifs. Also a **simplified Karnaugh map** could do the job, which less faster than look up table, but it takes much lesser ram space.  The point of having NN is to train them on a dataset and then predict unknown input. But it's fine to experiment with those stuff, so I give a credit to [Sean Hodgins] because learning is important."

--have fun ;-)

--or maybe someday even get an experiment sent to the ISS: https://code4space.org/

<img src="https://github.com/user-attachments/assets/f7c7432b-15e1-4172-a110-d7f67d30cc40" width="300">

___________________________________________________
**Note on network training:**
**augustineprem5587**
I face a problem while training the neural network. Error doesn't reach below the success rate, instead it goes back to training cycle 49 after going closer to the success threshold.. Can anyone help me?

**alexeykrylov9995**:
I don't know your particular situation, but you can start by playing with learning rate. In your case (validation and training errors change erratically and stay high), try to decrease learning rate.  Another possible problem is too small training set in relation to NN size (e.g., the number of weights)-- the symptoms are training error falling to nearly 0, while validation error starts by falling a little, and then grows back -- in such case make your NN smaller (or training set larger) .

**augustineprem5587**
alexeykrylov9995  Yeah! It worked.. Thanks man!!

**Reply by Growflavor** [reply]
**Testing the 'roachbot' in the ORL sim**...(ORL seems to have fixed locations for the sensors):

![RoachBotSim-ezgif com-video-to-gif-converter](https://github.com/user-attachments/assets/1ad93c90-d296-48ea-84af-4ba074d7ae52)

![Screenshot 2024-09-02 220114](https://github.com/user-attachments/assets/8c134c38-3d33-4016-b785-8b6aea74a7f2)

**1st evaluation on the EV3 "RoachBot"**...using reflected light mode...

<img src="https://github.com/user-attachments/assets/9ba471d0-35f9-4b2d-9ba7-d1b85d2b53a8"  width="300">

![Screenshot 2024-09-05 233955m](https://github.com/user-attachments/assets/40514baf-46e5-4bbb-a1e3-215c4f3e154e)

**Reply by Growflavor** [reply]
The 1st chase using 'ambient light' mode & a flashlight :-)
Now we can tinker with improving the xNN training...
Shared in the ORL gallery as "xNNroachBot".

<img src="https://github.com/user-attachments/assets/68c7e9a6-e13d-44b5-a520-e4159e828e28" height ="250">
<img src="https://github.com/user-attachments/assets/913d528c-8b96-40a3-baf6-ed099120d223" height ="250">

For comparison, here is the original project creator's video of his Arduino 'RoachBot': https://youtu.be/97R3TcUh5eI?si=pg-5JSxUJRD3hQWr&t=1036

<details>
  <summary>Training Data from seanhodgins' code formatted for upload to "Neural Network  Learn" in ORL:</summary>
  
ORL accepts a .txt file with:

_Input Nodes_; **OutputNodes**

_n1;n3;n4;n5_;**n2;n6**

format is separated by ; for training data:
  
  ```
0;1;1;0;0.65;0.55
0;1;0;0;0.75;0.5
1;1;1;0;0.2;0.2
1;1;0;0;1;0.2
0;0;1;0;0.5;0.75
1;0;0;0;0.3;0.3
0;0;0;0;0.5;0.5
0;0;0;1;0.75;0.75
0;1;0;1;1;0.75
0;0;1;1;0.75;1
0;1;1;1;1;1
1;0;0;1;1;0
1;1;0;1;1;0.75
1;0;1;1;0.75;1
1;0;1;0;0.2;1
1;1;1;1;0.65;0.65
  ```  
</details>
